File: ghidra_scripts/ExportDecompiled.java
// Ghidra script to export decompiled C code
// @category Export
// @author Claude Code

import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.util.task.ConsoleTaskMonitor;

import java.io.FileWriter;
import java.io.PrintWriter;

public class ExportDecompiled extends GhidraScript {
    @Override
    public void run() throws Exception {
        String programName = currentProgram.getName();
        String baseName = programName.replace(".cpython-39.so", "").replace(".so", "").replace(".o", "");
        String outputDir = "/home/gnydick/IdeaProjects/k2plus/decompiled/";
        String outputFile = outputDir + baseName + "_decompiled.c";

        DecompInterface decomp = new DecompInterface();
        decomp.openProgram(currentProgram);

        FunctionIterator funcIter = currentProgram.getFunctionManager().getFunctions(true);
        int funcCount = currentProgram.getFunctionManager().getFunctionCount();

        PrintWriter writer = new PrintWriter(new FileWriter(outputFile));
        writer.println("/*");
        writer.println(" * Decompiled from: " + programName);
        writer.println(" * Function count: " + funcCount);
        writer.println(" * Tool: Ghidra + Claude Code RE analysis");
        writer.println(" */\n");

        int processed = 0;
        while (funcIter.hasNext() && !monitor.isCancelled()) {
            Function func = funcIter.next();
            DecompileResults results = decomp.decompileFunction(func, 60, monitor);
            if (results != null && results.decompileCompleted()) {
                if (results.getDecompiledFunction() != null) {
                    writer.println("/* ============================================");
                    writer.println(" * Function: " + func.getName());
                    writer.printf(" * Address:  0x%x\n", func.getEntryPoint().getOffset());
                    writer.println(" * ============================================ */");
                    writer.println(results.getDecompiledFunction().getC());
                    writer.println("\n");
                }
            }
            processed++;
            if (processed % 100 == 0) {
                println("Processed " + processed + " / " + funcCount + " functions");
            }
        }
        writer.close();
        decomp.dispose();
        println("Exported: " + outputFile + " (" + funcCount + " functions)");
    }
}

File: ghidra_scripts/ExportDecompiled.py
# Ghidra script to export decompiled C code
# @category Export
# @author Claude Code

from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor
import os

def run():
    program_name = currentProgram.getName()
    base_name = program_name.replace(".cpython-39.so", "").replace(".so", "").replace(".o", "")
    output_dir = "/home/gnydick/IdeaProjects/k2plus/decompiled/"
    output_file = output_dir + base_name + "_decompiled.c"

    decomp = DecompInterface()
    decomp.openProgram(currentProgram)
    monitor = ConsoleTaskMonitor()

    func_manager = currentProgram.getFunctionManager()
    func_count = func_manager.getFunctionCount()

    with open(output_file, 'w') as f:
        f.write("/*\n")
        f.write(" * Decompiled from: %s\n" % program_name)
        f.write(" * Function count: %d\n" % func_count)
        f.write(" * Tool: Ghidra + Claude Code RE analysis\n")
        f.write(" */\n\n")

        processed = 0
        for func in func_manager.getFunctions(True):
            if monitor.isCancelled():
                break

            results = decomp.decompileFunction(func, 60, monitor)
            if results and results.decompileCompleted():
                decomp_func = results.getDecompiledFunction()
                if decomp_func:
                    f.write("/* ============================================\n")
                    f.write(" * Function: %s\n" % func.getName())
                    f.write(" * Address:  0x%x\n" % func.getEntryPoint().getOffset())
                    f.write(" * ============================================ */\n")
                    f.write(decomp_func.getC())
                    f.write("\n\n")

            processed += 1
            if processed % 100 == 0:
                print("Processed %d / %d functions" % (processed, func_count))

    print("Exported: %s (%d functions)" % (output_file, func_count))

run()
